### PR TITLE
Redirects with fragment

### PR DESCRIPTION
**Pull request by juunas11**
Added overloads to the methods that return redirect results that allow specifying the fragment to add to the URL.
 - Added RedirectToAction, RedirectToActionPermanent, RedirectToRoute, and RedirectToRoutePermanent overloads that allow specifying the fragment
 - Added the Fragment property to RedirectToActionResult and RedirectToRouteResult
 - Modified RedirectToActionResultExecutor and RedirectToRouteResultExecutor to call the overload on UrlHelper that specifies the fragment. Left host and protocol to null as the other UrlHelper overloads do.

Addresses issue #3988 

This is my first PR to MVC Core, I would be happy about any feedback :)

### CODE REVIEW

**Comment by Eilon** [comment]
@juunas11 thanks for the PR, we'll take a look!

@dougbu assigning to you to review and merge.

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs [hunk 1]
```diff
@@ -1082,7 +1219,7 @@ public virtual AcceptedAtRouteResult AcceptedAtRoute(object routeValues)
         /// <summary>
         /// Creates a <see cref="AcceptedAtRouteResult"/> object that produces an Accepted (202) response.
         /// </summary>
-        /// <param name="routeName">The name of the route to use for generating the URL.</param>       
+        /// <param name="routeName">The name of the route to use for generating the URL.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: 👏

File: src/Microsoft.AspNetCore.Mvc.Core/RedirectToActionResult.cs [hunk 2]
```diff
@@ -23,12 +23,32 @@ public class RedirectToActionResult : ActionResult, IKeepTempDataResult
             string actionName,
             string controllerName,
             object routeValues,
+            string fragment)
+            : this(actionName, controllerName, routeValues, permanent: false, fragment: fragment)
+        {
+        }
+
+        public RedirectToActionResult(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/RedirectToActionResult.cs]: Please add doc comments for this and the other new constructor overload. No need to do the same for the existing constructor overloads -- unless you're feeling industrious! 😺

File: src/Microsoft.AspNetCore.Mvc.Core/RedirectToActionResult.cs [hunk 3]
```diff
@@ -53,6 +73,11 @@ public class RedirectToActionResult : ActionResult, IKeepTempDataResult
 
         public bool Permanent { get; set; }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/RedirectToActionResult.cs]: If you're feeling particularly industrious, feel free to add doc comments for this property as well 😺

File: src/Microsoft.AspNetCore.Mvc.Core/RedirectToRouteResult.cs [hunk 4]
```diff
@@ -27,10 +27,28 @@ public RedirectToRouteResult(object routeValues)
             string routeName,
             object routeValues,
             bool permanent)
+            : this(routeName, routeValues, permanent, fragment: null)
+        {
+        }
+
+        public RedirectToRouteResult(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/RedirectToRouteResult.cs]: My doc comment and if-industrious comments in `RedirectToActionResult` apply here too.

File: src/Microsoft.AspNetCore.Mvc.Core/RedirectToRouteResult.cs [hunk 5]
```diff
@@ -50,6 +68,11 @@ public RedirectToRouteResult(object routeValues)
 
         public bool Permanent { get; set; }
 
+        /// <summary>
+        /// Gets or sets the fragment to be added to the URL
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/RedirectToRouteResult.cs]: "to be added" -> "to add"

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs [hunk 6]
```diff
@@ -469,9 +469,46 @@ public virtual RedirectToActionResult RedirectToAction(string actionName, string
...
+            return RedirectToAction(actionName, controllerName, routeValues, fragment: null);
+        }
+
+        /// <summary>
+        /// Redirects to the specified action using the specified <paramref name="actionName"/>,
+        /// <paramref name="controllerName"/>, and <paramref name="fragment"/>.
+        /// </summary>
+        /// <param name="actionName">The name of the action.</param>
+        /// <param name="controllerName">The name of the controller.</param>
+        /// <param name="fragment">The fragment to add to the URL</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: Please add a period at the end of this comment. Same for the `fragment` doc comment in the overload just below.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs [hunk 7]
```diff
@@ -83,6 +83,41 @@ public class RedirectToActionResultTest
                 "No route matches the supplied values.");
         }
 
+        [Fact]
+        public async Task RedirectToAction_Execute_WithFragment_PassesCorrectValuesToRedirect()
+        {
+            // Arrange
+            var expectedUrl = "/Home/SampleAction#test";
+            var expectedPermanentFlag = false;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs]: nit: just `expectedPermanent`

File: test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs [hunk 8]
```diff
@@ -83,6 +83,41 @@ public class RedirectToActionResultTest
...
+
+            var httpContext = new Mock<HttpContext>();
+            httpContext
+                .SetupGet(o => o.RequestServices)
+                .Returns(CreateServices().BuildServiceProvider());
+
+            var httpResponse = new Mock<HttpResponse>();
+            httpContext
+                .Setup(o => o.Response)
+                .Returns(httpResponse.Object);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs]: Should be no need to mock `HttpContext` or `HttpResponse`. Try using `DefaultHttpContext` and setting `RequestServices` instead.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs [hunk 9]
```diff
@@ -83,6 +83,41 @@ public class RedirectToActionResultTest
...
+            var result = new RedirectToActionResult("SampleAction", "Home", null, false, "test")
+            {
+                UrlHelper = urlHelper,
+            };
+
+            // Act
+            await result.ExecuteResultAsync(actionContext);
+
+            // Assert
+            // Verifying if Redirect was called with the specific Url and parameter flag.
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs]: nit: "the specific Url and parameter flag" -> "the expected URL and permanent parameters"

File: test/WebSites/BasicWebSite/Controllers/HomeController.cs [hunk 10]
```diff
@@ -35,11 +35,21 @@ public IActionResult RedirectToActionReturningTaskAction()
             return RedirectToAction("ActionReturningTask");
         }
 
+        public IActionResult RedirectToActionWithFragment()
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Controllers/HomeController.cs]: This and the method below are not used. Probably should be. I'm not saying we need tests of all 8 new `ControllerBase` overloads but zero is too low. Suggest covering the two methods used here and the longest `RedirectToActionPermanent()` overload.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs [hunk 11]
```diff
@@ -83,6 +83,41 @@ public class RedirectToActionResultTest
                 "No route matches the supplied values.");
         }
 
+        [Fact]
+        public async Task RedirectToAction_Execute_WithFragment_PassesCorrectValuesToRedirect()
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs]: Add a similar test in `RedirectToRouteResultTest`, perhaps covering the overload accepting `permanent: true` instead.

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs (hunk 1 above)
Comment (juunas11) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: I can undo this one if you want, left the whitespace deleter on auto :)

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs [hunk 12]
```diff
@@ -469,9 +469,46 @@ public virtual RedirectToActionResult RedirectToAction(string actionName, string
...
+        public virtual RedirectToActionResult RedirectToAction(
+            string actionName,
+            string controllerName,
+            object routeValues,
+            string fragment)
+        {
+            return new RedirectToActionResult(actionName, controllerName, routeValues, fragment)
             {
-                UrlHelper = Url,
+                UrlHelper = Url
```
Comment (juunas11) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: Did not know it was a style decision :) Thanks for pointing out, will do so in the future.

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: I was thanking you here. Removing trailing whitespace in a file you're changing is fine.

File: src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs [hunk 13]
```diff
@@ -610,7 +716,38 @@ public virtual RedirectToRouteResult RedirectToRoutePermanent(object routeValues
         [NonAction]
         public virtual RedirectToRouteResult RedirectToRoutePermanent(string routeName, object routeValues)
         {
-            return new RedirectToRouteResult(routeName, routeValues, permanent: true)
+            return RedirectToRoutePermanent(routeName, routeValues, fragment: null);
+        }
+
+        /// <summary>
+        /// Redirects to the specified route with <see cref="RedirectToRouteResult.Permanent"/> set to true
+        /// using the specified <paramref name="routeName"/>, and <paramref name="fragment"/>.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs]: Remove the comma here.

File: src/Microsoft.AspNetCore.Mvc.Core/RedirectToActionResult.cs [hunk 14]
```diff
@@ -51,8 +108,16 @@ public class RedirectToActionResult : ActionResult, IKeepTempDataResult
         /// </summary>
         public RouteValueDictionary RouteValues { get; set; }
 
+        /// <summary>
+        /// Gets or sets whether the redirect is permanent.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/RedirectToActionResult.cs]: "Gets or sets an indication that &hellip;"

File: src/Microsoft.AspNetCore.Mvc.Core/RedirectToRouteResult.cs [hunk 15]
```diff
@@ -48,8 +104,16 @@ public RedirectToRouteResult(object routeValues)
         /// </summary>
         public RouteValueDictionary RouteValues { get; set; }
 
+        /// <summary>
+        /// Gets or sets whether the redirect is permanent.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/RedirectToRouteResult.cs]: "Gets or sets an indication that &hellip;"

File: test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs (hunk 8 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/RedirectToActionResultTest.cs]: Guess the existing tests weren't the best model to follow. Don't change them but do avoid unnecessary mocks in tests you add.

**Comment by dougbu** [comment]
@juunas11 this PR is very close. Any chance you can wrap up the few remaining comments soon?

**Comment by juunas11** [comment]
Yes, I'll try to get everything done in the following one or two days.

**Comment by juunas11** [comment]
I *think* I've done everything you requested, but please tell me if there is still something.

File: test/WebSites/BasicWebSite/Controllers/HomeController.cs (hunk 10 above)
Comment (dougbu) [inline test/WebSites/BasicWebSite/Controllers/HomeController.cs]: Please revert the changes to this test controller. None of the functional tests uses the two new methods and I've decided coverage is fine with the new unit tests.

**Comment by dougbu** [comment]
1dd1d49
